### PR TITLE
dkg/disk: fetch definition file from remote uri

### PR DESCRIPTION
**Pull request by xenowits**
Add support to fetch `cluster-definition` file from a remote URI.

category: feature
ticket: #1045

### CODE REVIEW

File: dkg/disk.go [hunk 1]
```diff
@@ -53,6 +61,27 @@ func loadDefinition(conf Config) (cluster.Definition, error) {
 	return res, nil
 }
 
+// fetchDefinition fetches cluster definition file from a remote URI.
+func fetchDefinition(url string) (cluster.Definition, error) {
+	resp, err := http.Get(url) //nolint:gosec,noctx
```
Comment (xenowits) [inline dkg/disk.go:66]: discuss: should i pass a `http.Client` as a function parameter?

File: dkg/disk.go [hunk 2]
```diff
@@ -53,6 +61,27 @@ func loadDefinition(conf Config) (cluster.Definition, error) {
...
+	}
+	defer resp.Body.Close()
+
+	buf, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return cluster.Definition{}, errors.Wrap(err, "read response body")
+	}
+
+	def := new(cluster.Definition)
+	if err := def.UnmarshalJSON(buf); err != nil {
```
Comment (corverroos) [inline dkg/disk.go:78]: rather do standard: `json.Unmarshal`

File: dkg/disk.go [hunk 3]
```diff
@@ -53,6 +61,27 @@ func loadDefinition(conf Config) (cluster.Definition, error) {
...
+		return cluster.Definition{}, errors.Wrap(err, "fetch file")
+	}
+	defer resp.Body.Close()
+
+	buf, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return cluster.Definition{}, errors.Wrap(err, "read response body")
+	}
+
+	def := new(cluster.Definition)
```
Comment (corverroos) [inline dkg/disk.go:77]: nit: `var def cluster.Definition` then `json.Unmarshal(data, &def)`

File: dkg/disk.go [hunk 4]
```diff
@@ -33,12 +36,17 @@ import (
 	"github.com/obolnetwork/charon/tbls/tblsconv"
 )
 
-// loadDefinition returns the cluster definition from disk (or the test definition if configured).
+// loadDefinition returns the cluster definition from disk or network. It returns the test definition if configured.
 func loadDefinition(conf Config) (cluster.Definition, error) {
```
Comment (corverroos) [inline dkg/disk.go:40]: add an internal test for this function please

File: dkg/disk.go [hunk 5]
```diff
@@ -33,12 +36,17 @@ import (
 	"github.com/obolnetwork/charon/tbls/tblsconv"
 )
 
-// loadDefinition returns the cluster definition from disk (or the test definition if configured).
+// loadDefinition returns the cluster definition from disk or network. It returns the test definition if configured.
```
Comment (corverroos) [inline dkg/disk.go:39]: also please update the flag description

Comment (corverroos) [inline dkg/disk.go:39]: `from disk or a HTTP URL`

File: dkg/disk.go [hunk 6]
```diff
@@ -33,12 +36,17 @@ import (
...
 func loadDefinition(conf Config) (cluster.Definition, error) {
 	if conf.TestDef != nil {
 		return *conf.TestDef, nil
 	}
 
+	_, err := url.ParseRequestURI(conf.DefFile)
```
Comment (corverroos) [inline dkg/disk.go:45]: suggest checking scheme for http or https

**Review by corverroos — APPROVED** [review]

File: dkg/disk.go (hunk 6 above)
Comment (corverroos) [reply dkg/disk.go:45]: what happens for `file://local/definition.jon`?

File: dkg/disk.go (hunk 1 above)
Comment (corverroos) [reply dkg/disk.go:66]: just pass in context and use a 5s timeout

File: dkg/disk.go (hunk 6 above)
Comment (xenowits) [reply dkg/disk.go:45]: it should be a valid URI with either`HTTP` or `HTTPS` scheme. It rejects other schemes like `file` or `ftp`

File: dkg/disk.go (hunk 4 above)
Comment (xenowits) [reply dkg/disk.go:40]: done. for both `loadDefinition` and `fetchDefinition`

File: dkg/disk.go (hunk 1 above)
Comment (corverroos) [reply dkg/disk.go]:
```
func fetchDefinition(ctx context.Context, url string) (cluster.Definition, error) {
    ctx, cancel := context.WithCancel(ctx, time.Second*5)
    defer cancel()
    
    req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
    
    resp, err = new(http.Client).Do(req)
```

File: dkg/disk_internal_test.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+			b, _ := validDef.MarshalJSON()
+			_, _ = w.Write(b)
+		case "/invalidDef":
+			b, _ := invalidDef.MarshalJSON()
+			_, _ = w.Write(b)
+		}
+	}))
+	defer server.Close()
+
+	type args struct {
```
Comment (corverroos) [inline dkg/disk_internal_test.go:51]: no need to this type, can just make it a string

File: dkg/disk_internal_test.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := fetchDefinition(context.Background(), tt.args.url)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("fetchDefinition() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
```
Comment (corverroos) [inline dkg/disk_internal_test.go:80]: use `require.Equal`

File: dkg/disk_internal_test.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	invalidFile := "invalid-cluster-definition.json"
+	err = os.WriteFile(invalidFile, []byte{1, 2, 3}, 0o666)
+	require.NoError(t, err)
+
+	defer func() {
+		require.NoError(t, os.Remove(validFile))
+		require.NoError(t, os.Remove(invalidFile))
+	}()
+
+	type args struct {
```
Comment (corverroos) [inline dkg/disk_internal_test.go:108]: same, just make it a string (DefFile values)

File: dkg/disk_internal_test.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+			name:    "Fetch invalid definition",
+			args:    args{url: fmt.Sprintf("%s/%s", server.URL, "invalidDef")},
+			want:    invalidDef,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := fetchDefinition(context.Background(), tt.args.url)
+			if (err != nil) != tt.wantErr {
```
Comment (corverroos) [inline dkg/disk_internal_test.go:76]: nit: use `require` package
```
if test.wantErr {
    require.Error(t, err)
    return
}
```

**Pull request merged**